src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,712 @@
+            }
+        }
+
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+
+        /// <inheritdoc />
+        void ICollection<KeyValuePair<TKey, TValue>>.Add(KeyValuePair<TKey, TValue> item)
+        {
+            if (_arrayStorage != null)
+            {
+                Add(item.Key, item.Value);
+                return;
+            }
+
+            ((ICollection<KeyValuePair<TKey, TValue>>)_dictionaryStorage!).Add(item);
+            return;
+        }
+
+        /// <inheritdoc />
+        public void Add(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_arrayStorage != null)
+            {
+                EnsureCapacity(_count + 1);
+
+                if (_dictionaryStorage != null)
+                {
+                    Debug.Assert(_arrayStorage == null);
+                    _dictionaryStorage.Add(key, value);
+                    return;
+                }
+
+                if (ContainsKeyArray(key))
+                {
+                    throw new ArgumentException($"An element with the key '{key}' already exists in the {nameof(AdaptiveCapacityDictionary<TKey, TValue>)}.", nameof(key));
+                }
+
+                _arrayStorage[_count] = new KeyValuePair<TKey, TValue>(key, value);
+                _count++;
+                return;
+            }
+
+            _dictionaryStorage!.Add(key, value);
+            return;
+        }
+
+        /// <inheritdoc />
+        public void Clear()
+        {
+            if (_dictionaryStorage != null)
+            {
+                _dictionaryStorage.Clear();
+            }
+
+            if (_count == 0)
+            {
+                return;
+            }
+            if (_arrayStorage != null)
+            {
+                Array.Clear(_arrayStorage, 0, _count);
+                _count = 0;
+            }
+        }
+
+        /// <inheritdoc />
+        bool ICollection<KeyValuePair<TKey, TValue>>.Contains(KeyValuePair<TKey, TValue> item)
+        {
+            if (_dictionaryStorage != null)
+            {
+                return ((ICollection<KeyValuePair<TKey, TValue>>)_dictionaryStorage).Contains(item);
+            }
+
+            return TryGetValue(item.Key, out var value) && EqualityComparer<object>.Default.Equals(value, item.Value);
+        }
+
+        /// <inheritdoc />
+        public bool ContainsKey(TKey key)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_dictionaryStorage != null)
+            {
+                return _dictionaryStorage.ContainsKey(key);
+            }
+
+            return ContainsKeyCore(key);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ContainsKeyCore(TKey key)
+        {
+            return ContainsKeyArray(key);
```

Comment:
Should the logic for `ContainsKeyArray` go here or remove `ContainsKeyCore`?

It's just about putting less work to the JIT and it's inliner (budget). `Contains` will inline in the code-path I checked, but with a deeper call-hierarchy this might hit JIT's inlining limit (I don't know the current limit from top of my head).